src/codegen/program.lisp

```diff
@@ -41,41 +42,96 @@ A function bound here will be called with a keyword category, and one or more ad
 
     Toplevel definitions, after type checking and before compilation.")
 
+;; The following functions control the output order of compiled
+;; definitions and interleaved lisp expressions.
+;;
+;; Toplevel define and instance forms are compiled to 1 or more named,
+;; lisp-source-valued output definitions: when these definitions are
+;; generated, they are associated with the starting source offset of their
+;; toplevel form:
+;;
+;;   toplevel definition:
+;;     #<def .... (300 . 345)>
+;;
+;;   bindings (lisp definitions):
+;;     (300 b1 .. bn)
+;;
+;; Then when compile-definitions emits the full set of output
+;; definitions, any lisp source forms that occurred earlier in the
+;; file are emitted first.
+
+(defun bindings-offset (bindings offsets)
+  "Given a list of binding names, and a name -> offset map, return the earliest binding start offset."
+  (apply #'min
+         (mapcar (lambda (binding)
+                   (or (gethash (car binding) offsets) 0))
+                 bindings)))
+
+(defun merge-forms (lisp defs)
+  "Merge a list of toplevel-lisp forms and compiled definitions, ensuring that lisp forms are always placed before the original source locations of the compiled definitions."
+  (let (result)
+    (loop :for def :in defs
+          :for next-def-offset := (car def)
+          :do (loop :while (< (caar lisp) next-def-offset)
+                    :do (push (pop lisp) result))
+              (push def result)
+          :finally (loop :while lisp
+                         :do (push (pop lisp) result)))
+    (nreverse result)))
+
+(defun compile-definitions (sccs definitions lisp-forms offsets env)
+  "Compile SCCs and generate a final output definition list, merging any present lisp sources."
+  (let ((bindings (loop :for scc :in sccs
+                        :for bindings := (remove-if-not (lambda (binding)
+                                                          (find (car binding) scc))
+                                                        definitions)
+                        :collect (cons (bindings-offset bindings offsets)
+                                       (compile-scc bindings env))))
+        (lisp-forms (mapcar (lambda (lisp-form)
+                              (cons (car (parser:toplevel-lisp-form-source lisp-form))
+                                    (parser:toplevel-lisp-form-body lisp-form)))
+                            lisp-forms)))
+    (mapcan #'cdr (merge-forms bindings lisp-forms))))
+
+(defun definition-bindings (definitions env offsets)
+  (loop :for define :in definitions
+        :for offset := (car (tc:toplevel-define-source define))
+        :for name := (tc:node-variable-name (tc:toplevel-define-name define))
+        :for compiled-node := (translate-toplevel define env)
+
+        :do (when *codegen-hook*
```

Comment:
nit. just out of curiosity, why not `:when *codegen-hook* :do (funcalll *codegen-hook* ...)`? Not a suggestion, the choice just jumped out at me.

Lean in to the `loop`... become the `loop` :P

Reply:
done